packages/router/src/active-route-loader.tsx

```diff
@@ -0,0 +1,260 @@
+  children?: React.ReactNode
+}
+
+const ArlNullPage = () => null
+const ArlWhileLoadingNullPage = () => null
+
+export const ActiveRouteLoader = ({
+  path,
+  spec,
+  delay,
+  params,
+  whileLoadingPage,
+  children,
+}: Props) => {
+  const location = useLocation()
+  const [pageName, setPageName] = useState('')
+  const loadingTimeout = useRef<NodeJS.Timeout>()
+  const announcementRef = useRef<HTMLDivElement>(null)
+  const mounted = useRef(true)
+  const waitingFor = useRef<string>('')
+  const [loadingState, setLoadingState] = useState<LoadingStateRecord>({
+    [path]: { page: ArlNullPage, state: 'PRE_SHOW' },
+  })
+  const [renderedChildren, setRenderedChildren] = useState<
+    React.ReactNode | undefined
+  >(children)
+  const [renderedPath, setRenderedPath] = useState(path)
+  const [renderedLocation, setRenderedLocation] = useState({ ...location })
+
+  loadingState[path] = loadingState[path] || 'PRE_SHOW'
+
+  const clearLoadingTimeout = () => {
+    if (loadingTimeout.current) {
+      clearTimeout(loadingTimeout.current)
+    }
+  }
+
+  useEffect(() => {
+    mounted.current = true
+
+    return () => {
+      mounted.current = false
+    }
+  }, [])
+
+  useEffect(() => {
+    global?.scrollTo(0, 0)
+
+    if (announcementRef.current) {
+      announcementRef.current.innerText = getAnnouncement()
+    }
+
+    const routeFocus = getFocus()
+    if (!routeFocus) {
+      resetFocus()
+    } else {
+      routeFocus.focus()
+    }
+  }, [pageName, params])
+
+  useEffect(() => {
+    const startPageLoadTransition = async (
+      { loader, name }: Spec,
+      delay: number = DEFAULT_PAGE_LOADING_DELAY
+    ) => {
+      setLoadingState((loadingState) => ({
+        ...loadingState,
+        [path]: {
+          page: ArlNullPage,
+          state: 'PRE_SHOW',
+        },
+      }))
+
+      // Update the context if importing the page is taking longer
+      // than `delay`.
+      // Consumers of the context can show a loading indicator
+      // to signal to the user that something is happening.
+      loadingTimeout.current = setTimeout(() => {
+        unstable_batchedUpdates(() => {
+          setLoadingState((loadingState) => ({
+            ...loadingState,
+            [path]: {
+              page: whileLoadingPage || ArlWhileLoadingNullPage,
+              state: 'SHOW_LOADING',
+            },
+          }))
+          setRenderedChildren(children)
+          setRenderedPath(path)
+          setRenderedLocation({ ...location })
+        })
+      }, delay)
+
+      // Wait to download and parse the page.
+      waitingFor.current = name
+      const module = await loader()
+
+      // Remove the timeout because the page has loaded.
+      clearLoadingTimeout()
+
+      // Only update all state if we're still interested (i.e. we're still
+      // waiting for the page that just finished loading)
+      if (mounted.current && name === waitingFor.current) {
+        unstable_batchedUpdates(() => {
```

Comment:
Just noting that this is stable in react 18:
- https://github.com/reactwg/react-18/discussions/21